<a href="https://colab.research.google.com/github/LokeRuiKee/ml-dl-playground/blob/main/DNN_experiment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DNN Template code

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='relu', input_shape=(784,)),  # Input layer with 128 neurons and ReLU activation
    Dense(64, activation='relu'),                                          # Hidden layer with 64 neurons and ReLU activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])


In [ ]:
# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)


11490434/11490434 [==============================] - 1s 0us/step
Epoch 1/5
1875/1875 [==============================] - 11s 5ms/step - loss: 0.2469 - accuracy: 0.9285
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1062 - accuracy: 0.9684
Epoch 3/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0727 - accuracy: 0.9778
Epoch 4/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0559 - accuracy: 0.9822
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0732 - accuracy: 0.9771
Test accuracy: 0.9771000146865845


# 7 activation, optimizer = adam

## relu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='relu', input_shape=(784,)),  # Input layer with 128 neurons and relu activation
    Dense(64, activation='relu'),                                          # Hidden layer with 64 neurons and relu activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.2395 - accuracy: 0.9296
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1020 - accuracy: 0.9687
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0707 - accuracy: 0.9783
Epoch 4/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0553 - accuracy: 0.9824
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0805 - accuracy: 0.9755
Test accuracy: 0.9754999876022339


## LeakyReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='LeakyReLU', input_shape=(784,)),  # Input layer with 128 neurons and LeakyReLU activation
    Dense(64, activation='LeakyReLU'),                                          # Hidden layer with 64 neurons and LeakyReLU activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 7s 3ms/step - loss: 0.2693 - accuracy: 0.9223
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1327 - accuracy: 0.9603
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0980 - accuracy: 0.9698
Epoch 4/5
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0794 - accuracy: 0.9744
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.1066 - accuracy: 0.9698
Test accuracy: 0.9697999954223633


## elu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='elu', input_shape=(784,)),  # Input layer with 128 neurons and elu activation
    Dense(64, activation='elu'),                                          # Hidden layer with 64 neurons and elu activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.2561 - accuracy: 0.9252
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1172 - accuracy: 0.9640
Epoch 3/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0805 - accuracy: 0.9747
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0604 - accuracy: 0.9810
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0843 - accuracy: 0.9745
Test accuracy: 0.9745000004768372


## PReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='PReLU', input_shape=(784,)),  # Input layer with 128 neurons and PReLU activation
    Dense(64, activation='PReLU'),                                          # Hidden layer with 64 neurons and PReLU activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 9s 4ms/step - loss: 0.2407 - accuracy: 0.9291
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1019 - accuracy: 0.9689
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0690 - accuracy: 0.9788
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0507 - accuracy: 0.9843
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0969 - accuracy: 0.9741
Test accuracy: 0.9740999937057495


## Sigmoid

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='sigmoid', input_shape=(784,)),  # Input layer with 128 neurons and sigmoid activation
    Dense(64, activation='sigmoid'),                                          # Hidden layer with 64 neurons and sigmoid activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 9s 4ms/step - loss: 0.4720 - accuracy: 0.8834
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1787 - accuracy: 0.9471
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1241 - accuracy: 0.9633
Epoch 4/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0932 - accuracy: 0.9721
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0870 - accuracy: 0.9726
Test accuracy: 0.972599983215332


## tanh

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.2616 - accuracy: 0.9245
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1193 - accuracy: 0.9640
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0798 - accuracy: 0.9762
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0591 - accuracy: 0.9822
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.0708 - accuracy: 0.9778
Test accuracy: 0.9778000116348267


## softmax

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='softmax', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='softmax'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 1.6436 - accuracy: 0.6841
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.6323 - accuracy: 0.8456
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.4599 - accuracy: 0.8788
Epoch 4/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.3777 - accuracy: 0.9078
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.3137 - accuracy: 0.9220
Test accuracy: 0.921999990940094


## conclusion
Best activation function = tanh . Test accuracy: 0.9778000116348267

# Best activation function: tanh, 8 optimizer

## SGD

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='SGD',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6023 - accuracy: 0.8466
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.3237 - accuracy: 0.9090
Epoch 3/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.2755 - accuracy: 0.9209
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2448 - accuracy: 0.9291
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.2073 - accuracy: 0.9414
Test accuracy: 0.9413999915122986


## SGD with Momentum

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer=SGDm,
              loss='sparse_categorical_crossentropy',  # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2956 - accuracy: 0.9135
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1451 - accuracy: 0.9570
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1051 - accuracy: 0.9680
Epoch 4/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0804 - accuracy: 0.9757
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0838 - accuracy: 0.9738
Test accuracy: 0.973800003528595


## RMSProp

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='RMSProp',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2614 - accuracy: 0.9243
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1247 - accuracy: 0.9623
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0869 - accuracy: 0.9740
Epoch 4/5
1875/1875 [==============================] - 7s 3ms/step - loss: 0.0665 - accuracy: 0.9797
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0795 - accuracy: 0.9756
Test accuracy: 0.975600004196167


## adam

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.2620 - accuracy: 0.9224
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1193 - accuracy: 0.9646
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0802 - accuracy: 0.9756
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0586 - accuracy: 0.9816
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0868 - accuracy: 0.9740
Test accuracy: 0.9739999771118164


## adamax

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adamax',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 7s 3ms/step - loss: 0.3509 - accuracy: 0.9034
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1931 - accuracy: 0.9445
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1459 - accuracy: 0.9574
Epoch 4/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1175 - accuracy: 0.9660
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.1023 - accuracy: 0.9678
Test accuracy: 0.9678000211715698


## nadam

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='nadam',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 9s 4ms/step - loss: 0.2615 - accuracy: 0.9247
Epoch 2/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1199 - accuracy: 0.9641
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0805 - accuracy: 0.9756
Epoch 4/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0602 - accuracy: 0.9808
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0854 - accuracy: 0.9736
Test accuracy: 0.9735999703407288


## AdaGrad

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='AdaGrad',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 1.0631 - accuracy: 0.7508
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.5773 - accuracy: 0.8634
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.4749 - accuracy: 0.8817
Epoch 4/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.4246 - accuracy: 0.8904
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.3675 - accuracy: 0.9041
Test accuracy: 0.9041000008583069


## AdaDelta

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='AdaDelta',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 8s 4ms/step - loss: 2.1853 - accuracy: 0.2245
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 1.9132 - accuracy: 0.4396
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 1.6886 - accuracy: 0.5749
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 1.5036 - accuracy: 0.6551
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 1.2709 - accuracy: 0.7279
Test accuracy: 0.7279000282287598


## conclusion
Best optimizer = RMSProp . Test accuracy: 0.975600004196167

# Experimenting Best Output Activation Function

Using Best activation function: tanh and Best optimizer: RMSProp

## output relu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='relu')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='RMSProp',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.3201 - accuracy: 0.1210
Epoch 2/5
1875/1875 [==============================] - 8s 4ms/step - loss: 2.2733 - accuracy: 0.1122
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.2678 - accuracy: 0.1142
Epoch 4/5
1875/1875 [==============================] - 7s 4ms/step - loss: 2.3013 - accuracy: 0.0993
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 2.2646 - accuracy: 0.1145
Test accuracy: 0.1145000010728836


## output  LeakyReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='LeakyReLU')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='RMSProp',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.3326 - accuracy: 0.2416
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.3025 - accuracy: 0.2307
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 2.3025 - accuracy: 0.2307
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.3025 - accuracy: 0.2307
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 2.3026 - accuracy: 0.2298
Test accuracy: 0.2298000007867813


## output elu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='elu')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='RMSProp',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 7s 3ms/step - loss: 2.3547 - accuracy: 0.2643
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 2.3024 - accuracy: 0.2432
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 2.3024 - accuracy: 0.2432
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 2.3024 - accuracy: 0.2433
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 2.3026 - accuracy: 0.2328
Test accuracy: 0.23280000686645508


## output PReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='PReLU')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='RMSProp',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.3414 - accuracy: 0.1006
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.3018 - accuracy: 0.0991
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.3026 - accuracy: 0.0987
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 2.3026 - accuracy: 0.0987
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 2.3026 - accuracy: 0.0980
Test accuracy: 0.09799999743700027


## output sigmoid

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='sigmoid')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='RMSProp',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2575 - accuracy: 0.9254
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1207 - accuracy: 0.9628
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0845 - accuracy: 0.9740
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0644 - accuracy: 0.9808
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0685 - accuracy: 0.9778
Test accuracy: 0.9778000116348267


## output tanh

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='tanh')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='RMSProp',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.3083 - accuracy: 0.2997
Epoch 2/5
1875/1875 [==============================] - 7s 3ms/step - loss: 2.3025 - accuracy: 0.2860
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 2.3025 - accuracy: 0.2875
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.3025 - accuracy: 0.2875
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 2.3026 - accuracy: 0.2859
Test accuracy: 0.2858999967575073


## output softmax

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Define the architecture of the DNN
model = Sequential([
    Dense(128, activation='tanh', input_shape=(784,)),  # Input layer with 128 neurons and tanh activation
    Dense(64, activation='tanh'),                                          # Hidden layer with 64 neurons and tanh activation
    Dense(10, activation='softmax')                                    # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='RMSProp',
              loss='sparse_categorical_crossentropy',           # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Preprocess the data
x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize pixel values to [0, 1]
# Flatten the data
x_train_flattened = x_train.reshape(-1, 28*28)
x_test_flattened = x_test.reshape(-1, 28*28)
# Train the model
model.fit(x_train_flattened, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test_flattened, y_test)
print('Test accuracy:', test_acc)

Epoch 1/5
1875/1875 [==============================] - 8s 3ms/step - loss: 0.2598 - accuracy: 0.9246
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1229 - accuracy: 0.9633
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0852 - accuracy: 0.9745
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0657 - accuracy: 0.9801
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0931 - accuracy: 0.9725
Test accuracy: 0.9725000262260437


## conclusion
Best output activation function: sigmoid. Test accuracy: 0.9778000116348267